### PR TITLE
Proxy forward auth data

### PR DESCRIPTION
**Pull request by jai1**
======

Currently the proxy extracts and forwards the client Principal to the broker. Client Principal is a modifiable string i.e can be changed or manufactured by the Proxy.

What we want to do instead is to send the clientAuthData to the broker - which is in most cases digitally signed. The broker will extract the client principal from the clientAuthData and reauthenticate the client.

In order to enforce this behavior we have introduced two new flags:-
- authenticateOriginalAuthData on the broker 
- forwardClientAuthData on the Proxy

======

@msb-at-yahoo

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -385,8 +465,11 @@ protected void handleConnect(CommandConnect connect) {
                 if (sslHandler != null) {
                     sslSession = ((SslHandler) sslHandler).engine().getSession();
                 }
-
-                originalPrincipal = connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null;
+                originalPrincipal = getOriginalPrincipal(
+                        connect.hasOriginalAuthData() ? connect.getOriginalAuthData() : null,
+                        connect.hasOriginalAuthMethod() ? connect.getOriginalAuthMethod() : null,
+                        connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null,
+                        sslSession);
```
Comment (saandrews) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java]: What happens if sslSession is null ?

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 2]
```diff
@@ -148,6 +148,13 @@ message CommandConnect {
 	// a Pulsar proxy. In this case the auth info above
 	// will the the auth of the proxy itself
 	optional string original_principal = 7;
+	
+	// Original auth role and auth Method that was passed 
+	// to the proxy. In this case the auth info above
+	// will the the auth of the proxy itself
```
Comment (saandrews) [inline pulsar-common/src/main/proto/PulsarApi.proto]: "will be the"?

Same for the remaining comments as well.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (jai1) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java]: is sslSession is null AuthDataCommand.hasDataFromTls will return false

**Comment by jai1** [comment]
@saandrews - have addressed your comments 
@rdhabalia @merlimat - Please review this PR when you find time

**Comment by jai1** [comment]
retest this please

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 3]
```diff
@@ -213,8 +221,19 @@ protected void handleLookup(CommandLookupTopic lookup) {
 
         final Semaphore lookupSemaphore = service.getLookupRequestSemaphore();
         if (lookupSemaphore.tryAcquire()) {
-            final String originalPrincipal = lookup.hasOriginalPrincipal() ? lookup.getOriginalPrincipal()
-                    : this.originalPrincipal;
+            String originalPrincipal;
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java]: This check can be done before acquiring the semaphore

Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java]: As with similar changes, try to put common logic in a single method

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java [hunk 4]
```diff
@@ -59,7 +59,10 @@
     private Set<String> authenticationProviders = Sets.newTreeSet();
     // Enforce authorization
     private boolean authorizationEnabled = false;
-
+    // Forward client authData to Broker for re authorization
+    // make sure authentication is enabled for this to take effect
+    private boolean forwardAuthData = false;
```
Comment (merlimat) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java]: Should be added to `proxy.conf`

Comment (merlimat) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java]: Rename to something like `authorizationForwardCredentials` or similar

**Comment by jai1** [comment]
> As with similar changes, try to put common logic in a single method

@merlimat - Have addressed your comments

I feel the last commit (https://github.com/apache/incubator-pulsar/pull/1169/commits/f986185eb7e79bcd3fac360d159057f39eeafcc6) addresses your comment but makes the logic more difficult to understand

**Comment by jai1** [comment]
@merlimat @rdhabalia - can I merge this or do you want to take one last look.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 5]
```diff
@@ -215,10 +217,18 @@ protected void handleLookup(CommandLookupTopic lookup) {
...
         }
 
+        String originalPrincipal = validateOriginalPrincipal(
+                lookup.hasOriginalAuthData() ? lookup.getOriginalAuthData() : null,
+                lookup.hasOriginalAuthMethod() ? lookup.getOriginalAuthMethod() : null,
+                lookup.hasOriginalPrincipal() ? lookup.getOriginalPrincipal() : this.originalPrincipal, requestId,
+                lookup);
+        
+        if (authenticateOriginalAuthData && lookup.hasOriginalAuthData() && originalPrincipal == null) {
+            return;
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java]: this condition seems little tricky. if we fail to validate original principal then `validateOriginalPrincipal()` will send failure response to the client and then thread should not process anything further. but then with this condition there is a possibility to move forward and thread will do further processing if `authenticateOriginalAuthData && lookup.hasOriginalAuthData() =is false`.??

**Comment by jai1** [comment]
@rdhabalia - handled your comments - hope to get a +1 from you

**Pull request merged**